samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/StatisticsService.java

```diff
@@ -0,0 +1,611 @@
+                    continue;
+                }
+
+                if (!answerCorrect) {
+                    // Found incorrect answer, we can count it and break te loop
+                    hasIncorrectAnswer = true;
+                    incorrectResponses++;
+                    break;
+                } else if (hasIncorrectAnswer == null) {
+                    hasIncorrectAnswer = false;
+                }
+            }
+
+            // Even if all selected answers are correct, we also need to compare the count
+            // to know that
+            // all correct answers were selected
+            if (Boolean.FALSE.equals(hasIncorrectAnswer) && selectedAnswerCount == correctAnswerCount) {
+                correctResponses++;
+            }
+        }
+
+        long attemptedResponses = correctResponses + incorrectResponses;
+
+        return ItemStatistics.builder()
+                .attemptedResponses(attemptedResponses)
+                .correctResponses(correctResponses)
+                .incorrectResponses(incorrectResponses)
+                .blankResponses(blankResponses)
+                .calcDifficulty()
+                .build();
+    }
+
+    private ItemStatistics getItemStatisticsForExtendedMatchingItem(ItemDataIfc item, Set<ItemGradingData> gradingData,
+            Set<PublishedAnswer> answers) {
+
+        Map<Long, Map<Long, Set<ItemGradingData>>> groupedGradingData = gradingData.stream()
+                .collect(Collectors.groupingBy(ItemGradingData::getAssessmentGradingId,
+                        Collectors.groupingBy(ItemGradingData::getPublishedItemTextId, Collectors.toSet())));
+
+        Map<Long, PublishedAnswer> answerMap = answers.stream()
+                .collect(Collectors.toMap(PublishedAnswer::getId, Function.identity()));
+
+        Map<Long, ItemTextIfc> itemTextMap = item.getItemTextSet().stream()
+                .collect(Collectors.toMap(ItemTextIfc::getId, Function.identity()));
+
+        long requiredOptionsCount = 0;
+        long correctResponses = 0;
+        long incorrectResponses = 0;
+        long blankResponses = 0;
+
+        for (Map<Long, Set<ItemGradingData>> submissionItemGradingData : groupedGradingData.values()) {
+            // int selectedAnswerCount = submissionItemGradingData.size();
+
+            int correctOptions = 0;
+            boolean isBlank = false;
+
+            // One option of an EMI item can have multiple answers, so we need to iterate
+            // through the grading data by item text
+            submissionLoop:
+            for (Map.Entry<Long, Set<ItemGradingData>> submissionGradingDataEntry : submissionItemGradingData.entrySet()) {
+                Long itemTextId = submissionGradingDataEntry.getKey();
+                ItemTextIfc itemText = itemTextMap.get(itemTextId);
+
+                Set<ItemGradingData> itemTextGradingData = submissionGradingDataEntry.getValue();
+
+                requiredOptionsCount++;//= requiredAnswerCount;
```

Comment:
Is this comment needed or just a leftover?

Reply:
to remove